src/NServiceBus.Core/Transports/Learning/LearningTransportMessagePump.cs

```diff
@@ -0,0 +1,293 @@
+        }
+
+        public void Start(PushRuntimeSettings limitations)
+        {
+            maxConcurrency = limitations.MaxConcurrency;
+            concurrencyLimiter = new SemaphoreSlim(maxConcurrency);
+            cancellationTokenSource = new CancellationTokenSource();
+
+            cancellationToken = cancellationTokenSource.Token;
+
+            if (purgeOnStartup)
+            {
+                Array.ForEach(Directory.GetFiles(path), File.Delete);
+            }
+            messagePumpTask = Task.Run(ProcessMessages, cancellationToken);
+
+            delayedMessagePoller.Change(TimeSpan.FromSeconds(1), TimeSpan.FromMilliseconds(-1));
+        }
+
+        public async Task Stop()
+        {
+            cancellationTokenSource.Cancel();
+
+            await messagePumpTask
+                .ConfigureAwait(false);
+
+            while (concurrencyLimiter.CurrentCount != maxConcurrency)
+            {
+                await Task.Delay(50, CancellationToken.None)
+                    .ConfigureAwait(false);
+            }
+
+            concurrencyLimiter.Dispose();
+        }
+
+        [DebuggerNonUserCode]
+        async Task ProcessMessages()
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    await InnerProcessMessages()
+                        .ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // graceful shutdown
+                }
+                catch (Exception exception)
+                {
+                    Logger.Error("File Message pump failed", exception);
+                }
+            }
+        }
+
+        async Task InnerProcessMessages()
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var filesFound = false;
+
+                foreach (var filePath in Directory.EnumerateFiles(path, "*.*"))
+                {
+                    filesFound = true;
+
+                    var nativeMessageId = Path.GetFileNameWithoutExtension(filePath);
+
+                    var transaction = GetTransaction();
+
+                    transaction.BeginTransaction(filePath);
+
+                    await concurrencyLimiter.WaitAsync(cancellationToken)
+                        .ConfigureAwait(false);
+
+                    InnerProcessFile(transaction, nativeMessageId).Ignore();
+                }
+
+                if (!filesFound)
+                {
+                    await Task.Delay(10, cancellationToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        ILearningTransportTransaction GetTransaction()
+        {
+            if (transactionMode == TransportTransactionMode.None)
+            {
+                return new NoTransaction(path);
+            }
+            var immediateDispatch = transactionMode == TransportTransactionMode.ReceiveOnly;
```

Comment:
This confused me, ImmediateDispatch is something we configure on send operations. But here we are creating a transaction. Is it really OK to default to immediate dispatch with ReceiveOnly? Even receive only could still benefit from batched dispatch